### PR TITLE
Fix Hawaii disabled exemption formula

### LINKED ISSUE #3350: Fix Hawaii disabled exemption formula
In the 0.552.0 version of Policyengine-US, the Hawaii disabled exemption amount is incorrectly calculated.   The following simple test case illustrates the problem:
```
- name: Hawaii disabled exemption bug
  absolute_error_margin: 0.01
  period: 2021
  input:
    people:
      person1:
        age: 30
        employment_income: 48_000
    tax_units:
      tax_unit:
        members: [person1]
    households:
      household:
        members: [person1]
        state_fips: 15  # HI
  output:  # expected results from hand-calculation with N-15 form/instructions
    hi_regular_exemptions: 1_144
    hi_disabled_exemptions: 0     # actual output is 2_288
    hi_exemptions: 1_144          # actual output is 2_288
```
This test case produces the following results:
```
  Test 'Hawaii disabled exemption bug':
    hi_disabled_exemptions@2021: [2288.] differs from 0.0
              with an absolute margin [2288.] > 0.01
```

### PR DESCRIPTION
**Pull request by YichunRen**
Fixes #3350

### Summary
🚀🐛🌎

This pull request fixes the Hawaii disabled exemption formula in `state_tax.py` and updates the changelog entry for version 0.1.1. This improves the state tax calculations in PolicyEngine.

> _`PolicyEngine`:_
> _Fixing state tax formulas_
> _A minor version_

### Walkthrough
* Fix the Hawaii disabled exemption formula to use the correct amount and threshold ()

### CODE REVIEW

**Comment by YichunRen** [comment]
The original `hi_disabled_exempetion` returns a non-zero value when both head and spouse are not disabled. Now, `hi_disabled_exempetion` returns a non-zero value only when the filer (head) is disabled.

File: policyengine_us/tests/policy/baseline/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.yaml [hunk 1]
```diff
@@ -40,4 +40,24 @@
     age_head: 63
     age_spouse: 65
   output:
-    hi_disabled_exemptions: 3_432
+    hi_disabled_exemptions: 0
+
+- name: Hawaii disabled exemption bug
```
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.yaml:45]: change the name to describe the situation

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py [hunk 2]
```diff
@@ -40,4 +40,6 @@ def formula(tax_unit, period, parameters):
             disabled_exemption_non_disabled_spouse,
         )
 
-        return disabled_exemption_head + disabled_exemption_spouse
+        return (
+            disabled_exemption_head + disabled_exemption_spouse
+        ) * disabled_head
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:45]: they get zero if only the spouse is disabled?

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py (hunk 2 above)
Comment (YichunRen) [reply policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py]: Sorry, the tax form instruction does not mention the case that only the spouse is disabled, so I wrongly think the spouse has to be the filer to claim the additional disabled exemption. However, the law does mention that case. I rewrote the formula according to state law and added some extra test units. Now, it should be correct for all the cases.

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py [hunk 3]
```diff
@@ -27,17 +27,40 @@ def formula(tax_unit, period, parameters):
...
+        disabled_exemption_head = where(
+            disabled_head,
+            max_(
+                disabled_exemption_disabled_head,
+                disabled_exemption_non_disabled_head,
+            ),
             disabled_exemption_non_disabled_head,
         )
         # Same for spouse.
         disabled_spouse = tax_unit("spouse_is_disabled", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:39]:
```suggestion
        disabled_spouse = tax_unit("spouse_is_disabled", period).astype(int)
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py [hunk 4]
```diff
@@ -24,20 +24,41 @@ def formula(tax_unit, period, parameters):
...
+        # if filling status is not joint, the disabled_exemption_spouse should be zero
+        # The taxpayer shall not take additional exemptions with regard to the taxpayer’s disability.
+        filing_status = tax_unit("filing_status", period)
+        disabled_exemption_spouse = where(
+            filing_status == filing_status.possible_values.JOINT,
+            disabled_exemption_spouse,
+            0,
+        )
+        no_disabled_head_or_spouse = (disabled_head + disabled_spouse) == 0
+        return where(
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:60]: why is this necessary? won't both parts of the sum be zero in this case?

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py [hunk 5]
```diff
@@ -24,20 +24,41 @@ def formula(tax_unit, period, parameters):
...
+            disabled_spouse,
+            max_(
+                disabled_exemption_disabled_spouse,
+                disabled_exemption_non_disabled_spouse,
+            ),
             disabled_exemption_non_disabled_spouse,
         )
 
-        return disabled_exemption_head + disabled_exemption_spouse
+        # if filling status is not joint, the disabled_exemption_spouse should be zero
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:51]:
```suggestion
        # If filing status is not joint, the disabled_exemption_spouse should be zero
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:1]: Could you try to make this code more concise? Without having fully compared this against the law, it seems unnecessarily long and complex.

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py (hunk 4 above)
Comment (YichunRen) [reply policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:34]: The exemption for a disabled taxpayer will be non-zero if the spouse is disabled and they are filing jointly. Thus, the formula's straightforward logic calculates the disabled exemption when disabled or the regular exemption if not. This line ensures a zero total exemption when both the head and spouse are not disabled, correcting the previous error.

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py
Comment (YichunRen) [reply policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:1]: I am fixing an error, so I tried not to touch previous code. The new version is now half as long, and I'm hopeful that it functions effectively

Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:1]: I'd suggest breaking this up into:
- `hi_disabled_exemptions_eligible`: either head or spouse has a disability
- `hi_disabled_exemptions_person`: person-level exemption amount, limited to heads or spouses and equal to $7,000 * disabled + $1,040 * (1 + aged)
- `hi_disabled_exemptions`: uses `adds` and `defined_for` to compute the final amount

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py
Comment (YichunRen) [reply policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:1]: Done!

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions_person.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+        "https://files.hawaii.gov/tax/forms/2022/n11ins.pdf#page=20"
+    )
+    definition_period = YEAR
+    defined_for = "hi_disabled_exemptions_eligible_person"
+
+    def formula(person, period, parameters):
+        p = parameters(period).gov.states.hi.tax.income.exemptions
+        disabled = person("is_disabled", period)
+        aged = person("age", period) >= p.aged_threshold
+        return max_(disabled * p.disabled, p.base * (1 + aged))
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions_person.py:19]:
```suggestion
        # Filer can claim greater of the disabled or base exemption.
        disabled_exemption = disabled * p.disabled
        # Aged individuals get an extra base exemption.
        base_exemption = p.base * (1 + aged)
        return max_(disabled_exemption, base_exemption)
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions_eligible_person.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions_eligible_person.py:1]: shouldn't this be tax unit-level, based on whether either head or spouse is disabled?

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
please add an integration test for a case with only spouse being disabled, which i don't think the current implementation will pass

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py
Comment (MaxGhenis) [reply policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:1]: not quite what i suggested

**Comment by PavelMakarchuk** [comment]
> please add an integration test for a case with only spouse being disabled, which i don't think the current implementation will pass

I have re-written the logic quite a bit as we were duplicating the base amount logic 

Now we have a separate computation for the head and spouse `hi_head_or_spouse_exemptions` where we include the aged and disabled logic and a `hi_regular_exemptions` file which includes all the other exemptions 

this should capture it accurately,

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
Still doesn't look to capture this:
>If you claim the Disability
Exemption, you will not be able to claim the additional
exemptions for your children or other dependents, or for
being 65 or older. 

I'd also suggest integrating the regular exemptions into the main logic. If you must separate out dependents, put dependent in the name and list dependents.

**Comment by PavelMakarchuk** [comment]
> Still doesn't look to capture this:
> 
> > If you claim the Disability
> > Exemption, you will not be able to claim the additional
> > exemptions for your children or other dependents, or for
> > being 65 or older.
> 
> I'd also suggest integrating the regular exemptions into the main logic. If you must separate out dependents, put dependent in the name and list dependents.

I have left the dependent exemptions separate as it seems more readable. Also added integration tests for the cases above

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_exemptions_person.py
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_exemptions_person.py:1]: rename file

File: policyengine_us/tests/policy/baseline/gov/states/hi/tax/income/exemptions/hi_dependent_exemptions.yaml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+- name: Head disbabled, not eligible
```
Comment (MaxGhenis) [inline policyengine_us/tests/policy/baseline/gov/states/hi/tax/income/exemptions/hi_dependent_exemptions.yaml:1]:
```suggestion
- name: Head disabled, not eligible
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]
let's follow the law/form by creating variables for normal exemptions and disabled exemptions, and taking the greater in the final formula

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_exemptions.py [hunk 8]
```diff
@@ -13,7 +13,6 @@ class hi_exemptions(Variable):
     defined_for = StateCode.HI
 
     def formula(tax_unit, period, parameters):
-        return max_(
-            tax_unit("hi_regular_exemptions", period),
-            tax_unit("hi_disabled_exemptions", period),
-        )
+        reuglar_exemptions = tax_unit("hi_regular_exemptions", period)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_exemptions.py:16]:
```suggestion
        regular_exemptions = tax_unit("hi_regular_exemptions", period)
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py [hunk 9]
```diff
@@ -13,31 +13,22 @@ class hi_disabled_exemptions(Variable):
...
-        disabled_exemption_non_disabled_spouse = p.base * (1 + aged_spouse)
-        disabled_exemption_spouse = max_(
-            disabled_exemption_disabled_spouse,
-            disabled_exemption_non_disabled_spouse,
-        )
-
-        return disabled_exemption_head + disabled_exemption_spouse
+        # The tax unit can claim exemptions for each dependent
+        dependents = tax_unit("tax_unit_dependents", period)
+        # The the head or the spouse is disabled, the dependent exemption goes to 0
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:20]:
```suggestion
        # If the head or the spouse is disabled, they take the disabled rather than dependent exemption.
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

**Comment by MaxGhenis** [comment]
tests fail

**Comment by PavelMakarchuk** [comment]
> tests fail

typo, fixed

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_regular_exemptions.py [hunk 10]
```diff
@@ -13,15 +13,12 @@ class hi_regular_exemptions(Variable):
     defined_for = StateCode.HI
 
     def formula(tax_unit, period, parameters):
+        exemptions = tax_unit("exemptions_count", period)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_regular_exemptions.py:16]: Use any time this shows up to disambiguate count from amount
```suggestion
        exemptions_count = tax_unit("exemptions_count", period)
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_regular_exemptions.py [hunk 11]
```diff
@@ -13,15 +13,12 @@ class hi_regular_exemptions(Variable):
...
-        exemption_count = (
-            tax_unit("exemptions_count", period) + aged_head + aged_spouse
-        )
-        return p.base * exemption_count
+        # Aged heads and spouses get an extra base exemption.
+        person = tax_unit.members
+        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
+        aged = person("age", period) >= p.aged_threshold
+        aged_head_or_spouse = tax_unit.sum(aged & head_or_spouse)
+        total_exemptions_including_aged = exemptions + aged_head_or_spouse
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_regular_exemptions.py:23]:
```suggestion
        total_exemption_count_including_aged = exemptions + aged_head_or_spouse
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_regular_exemptions.py [hunk 12]
```diff
@@ -13,15 +13,12 @@ class hi_regular_exemptions(Variable):
...
-        ).astype(int)
-        exemption_count = (
-            tax_unit("exemptions_count", period) + aged_head + aged_spouse
-        )
-        return p.base * exemption_count
+        # Aged heads and spouses get an extra base exemption.
+        person = tax_unit.members
+        head_or_spouse = person("is_tax_unit_head_or_spouse", period)
+        aged = person("age", period) >= p.aged_threshold
+        aged_head_or_spouse = tax_unit.sum(aged & head_or_spouse)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_regular_exemptions.py:22]:
```suggestion
        aged_head_spouse_count = tax_unit.sum(aged & head_or_spouse)
```

File: policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py [hunk 13]
```diff
@@ -13,31 +13,32 @@ class hi_disabled_exemptions(Variable):
...
+        total_conditional_head_or_spouse_amount = tax_unit.sum(
+            conditional_head_or_spouse_amount
         )
-
-        return disabled_exemption_head + disabled_exemption_spouse
+        total_amount = (
+            conditional_dependent_amount
+            + total_conditional_head_or_spouse_amount
+        )
+        return where(any_disabled_head_or_spouse, total_amount, 0)
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/hi/tax/income/exemptions/hi_disabled_exemptions.py:44]:
```suggestion
        return any_disabled_head_or_spouse * total_amount
```

**Review by MaxGhenis — APPROVED** [review]
Good to merge after this. Please also resolve all outstanding comment threads.
